content/blog/2019-11-02-custom-alloc/index.md

```diff
@@ -0,0 +1,206 @@
+* It also prevents you from using a different custom allocator to do something else that's more valuable in the future
+* It keeps you from using garbage collection
+
+
+
+# The Paper: Reconsidering Custom Memory Management
+
+The paper at hand, [Reconsidering Custom Memory Allocation][paper], is packaged and sold as the thesis that custom memory allocators in practice do not out perform general ones. 
+
+One of the key arguments of this paper is that standard baselines are not fair. Evidently the usual argument _in favor of_ cusom allocation in 2002 was a comparison against the win32 memory allocator, which was much slower than custom allocators. The first part of this paper is an evaluation against the Lea allocator (Doug Lea's `dlmalloc`, now often the default allocator), which greatly reduces and in some cases eliminates the margin of victory for custom allocators not exploting regions. 
+
+
+### The Taxonomy of Memory Allocators.
+This paper uses the following three bins to classify allocators: 
+
+**[1] Per-class allocators**. You build a separate heap for every object size class, and optimize each one separately for objects of this size. This is fast, simple, and interacts with C well --- but could be space inefficient if you don't know how much memory of each class size you will use.
+
+
+**[2] Regions**.  An allocator that allocates large chunks of memory, puts smaller pieces within it, and then must free them all at once. They are fast and convenent, but use a lot of space and are arguably dangrous because dangling references keep things from being freed. This (nominally) requires programmers to re-structure code to keep references to the region, and free the entire region at once, resulting in a usage pattern that looks somewhat like this:
+
+	```
+	createRegion(r);
+	x1 = regionMalloc(r,8);
+	x2 = regionMalloc(r,8);
+	x3 = regionMalloc(r,16);
+	x4 = regionMalloc(r,8);
+	```
+	
+**[3] Custom patterns**. Anything else --- for example, those that exploit stack-like patterns in memory allocation (the relevant benchmark is `197.parser`). The authors describe these as fast, but brittle.
+
+## Experimental Setup
+
+Although this paper is well-regarded and remembered primarily for its evaluation, the actual experimental data involves only 8 programs, each run with a single input. The table containing the programs and their inputs can be seen in the figure below:
+
+<!-- ![](expts.png) -->
+<img src="expts.png" alt="drawing" width="400" style="margin-left:100px;"/>
+
+All of the other graphs in this blog post will reference the 8 benchmarks on the left, and run their experiments for a single input, a single time, not reporting variance, the impact of parameters, machine loads, or any other confounders. For a paper that is purported to be an exemplar takedown of a common design practice, and appears to have had a large impact on the state of memory allocation, its actual empirical backing is quite flimsy.
```

Comment:
Maybe something nicer like "could be more robust"? "Quite flimsy" seems somewhat needlessly harsh. Your point is plenty clear without it.